Use openssl 1.1 by default

globin opened the pull request:
Tracking hydra jobset: https://hydra.nixos.org/jobset/nixpkgs/openssl-1.1

globin committed:
openssl: default to openssl_1_1_0

globin committed:
postgresql: build < 9.5 with openssl 1.0.2

vcunat commented:
I guess this isn't yet ready to be staged; it seems many packages can't handle the new version yet.

globin commented:
Yes, still wip

globin committed:
pin some packages to openssl 1.0.2 for now

vcunat commented:
Perhaps the package updates are worth merging regardless of default openssl version?

globin commented:
Was planning to finish this in the next few days anyway so not sure if it's worth it.

globin committed:
uwimap: add patch to build with openssl 1.1.0

globin committed:
qca2: explicitly require openssl_1_0_2

globin commented:
Retargeting for after 17.03 as too much stuff is breaking and not easily fixable due to openssl being propagated sometimes and stuff (PHP) trying to link to both versions due to dependencies.

I'll merge the version bumps to staging.

vcunat commented:
BTW, #24574 claims to fix ntp
> * [Bug 3363] Support for openssl-1.1.0 without compatibility modes
>   - rework of patch set from <ntp.org@eroen.eu>. <perlinger@ntp.org>

globin committed:
libexosip: add patch for openssl 1.1

globin committed:
aircrack-ng: add unreleased patch for openssl 1.1

globin committed:
duo-unix: 1.9.19 -> 1.9.20

fixes build with openssl 1.1

vcunat commented:
Still far from finishing this monster? Perhaps some stuff would easy to get to staging or master directly? (like those various updates)

globin committed:
opendkim: add patch for openssl 1.1

globin committed:
eiskaltdcpp: 2.2.9 -> 2.2.10, add upstream openssl patches

globin committed:
kdeFrameworks.kdelibs4support: add patch for openssl 1.1

FRidh commented:
Considering the amount of failures is already quite low (basically at the noise level) it would seem to me this is ready to go in staging. What do you think @globin?

globin commented:
I'd prefer fixing up a but more, i.e. everything depending on qt 5.6 is broken..

FRidh commented:
@globin Qt 5.6 seems to be broken on staging as well.
https://nix-cache.s3.amazonaws.com/log/8hya81bnhab2b0mps9amld1ma989hz44-qtbase-5.6.2.drv

fpletz committed:
ima-evm-utils: build with openssl 1.0.2

fpletz committed:
chaps: mark as broken due to openssl 1.1

fpletz committed:
shellinabox: build with openssl 1.0.2

globin commented:
Mostly reminder to myself: erlang < R20 needs patching

vcunat commented:
I wonder, do you still plan to have this in 17.09?

fpletz commented:
Bumped the milestone to 18.03. We're not expecting this to be ready for 17.09.

copumpkin committed:
xar: fix to work on openssl 1.1.x

copumpkin commented:
`xar` should be happy again. I even checked the box! 😄

NeQuissimus commented:
I wonder how much we should really care about old Erlang and Ruby versions...

peti commented:
I'm currently updating this branch to fit on top of the current `master` @ b3970e7f5c2368fa9063aacbc2d6c99c62f78e9c. I'll push a new version in a couple of hours and report my findings.

peti commented:
I force-pushed a freshly rebased version of this branch (b3197666c83ab10e04f24e4cc0d3d87897e6787b) on top of the current `master`. The original branch contents is still available for reference at https://github.com/peti/nixpkgs/tree/openssl-1.1-pre-rebase (5791eff8d80c786d26837fc64980b19330001e22). I'll wait for Hydra to deliver a preliminary verdict on the current state of affairs.

vcunat commented:
Bigger breakages I can see ATM (some might only appear with updated master):
- [ ] subversion
- [ ] openssh &ndash; update conflicting with the huge openssl-1.1 patch

vcunat committed:
Revert "kdeFrameworks.kdelibs4support: add patch for openssl 1.1"

This reverts commit 19f770218e053cd9542ecffef0eac604707fce96.
It's no longer applicable nor needed.

vcunat committed:
serf: build with openssl-1.0.2 again

The linkage was bad.  It built but at load-time it demanded symbols that
don't exist in 1.1.0 and thus subversion wouldn't build.

vcunat commented:
Let me take the baton for this evening, then I'll make Hydra re-evaluate with my fixes, and we can hopefully merge soon-ish.

vcunat committed:
eiskaltdcpp: remove duplicate patches after merge

vcunat commented:
I now see `qt5*` as significant problems to solve.  They're probably not very good cases to override to different openssl version, due to being a frequent dependency. (I expect collisions if two versions got into one process.)

vcunat committed:
qt59.qtbase: fix with openssl-1.1.0 via ported Fedora patch

vcunat commented:
I didn't notice that the default 5.10 builds out of the box.  I patched 5.9 in d9e691822, and I expect 5.6 can be removed?  What maintainers think of the removal? @ttuegel, @qknight, @periklis.

vcunat commented:
We might be approaching a point where we could merge and leave the rest to the community &ndash; roughly 150 [build regressions on x86_64-linux](https://hydra.nixos.org/eval/1433051?filter=x86_64-linux&compare=1432644&full=#tabs-now-fail) (almost nothing tried on darwin yet).

peti commented:
> We might be approaching a point where we could merge and leave the rest to the community.

That's what I thought, too. The current effort is quite comprehensive already, and expecting the handful of contributors to this branch to fix *all of Nixpkgs* doesn't seem realistic.

vcunat commented:
@globin: do you remember how the **ruby** breakage manifests?  All the versions seem to build and respond to `--help` and `--version`, and I can't see any comment fixing it (except for switch to 2.4 by default).  Hidden breakages tend to be unpleasant.

vcunat commented:
The same goes for **erlang** (<R20?)

peti reviewed (approved):
I tested several builds locally and everything worked fine.

👍 for merging.

periklis commented:
@vcunat Can't we throw this PR against a darwin slave to check out the regressions?

peti commented:
@periklis, the Hydra job *does* build Darwin.

periklis commented:
Ok that's what i thought. So then @vcunat means, "nothing tried on Darwin" as "nothing tried on darwin on runtime", right?

vcunat commented:
No, Hydra's been rather overloaded on Darwin side, so it didn't get to building many of the builds.

vcunat commented:
ruby: @zimbatm, erlang: @gleber (from `.github/CODEOWNERS`).
tl;dr: older ruby/erlang versions possibly have problems with openssl-1.1.0.  What do you think?

NeQuissimus commented:
Why not just get rid of Ruby 2.3? 2.4 and 2.5 have been out for a bit...

vcunat commented:
@NeQuissimus: see https://github.com/NixOS/nixpkgs/pull/34927

globin commented:
Please do not merge this right now, I want to go through this again first and maybe clean up stuff.

globin commented:
Wrt to ruby at least, the older versions don't build the ssl extension as it "cannot find openssl" which doesn't break the build but the world of ruby libraries trying to import ssl.

zimbatm commented:
this PR is now conflicting, probably because of the ruby changes introduced in #34927

vcunat commented:
Ruby changes in; Hydra's working on updated binaries, so we can test more easily.  @zimbatm: ruby 2.3 is kept, so the preferred way is to make it use older openssl?  I suppose "we" don't like the current state in this branch: succeeding build with no openssl support.

zimbatm commented:
yes, are we keeping a older openssl around? setting the version for ruby 2.3 would be great.

vcunat commented:
Of course we are.  A dozen packages keep pinned to `1.0.2`.

globin commented:
We have to be very(!) careful with pinning stuff to 1.0.2!! If something is linked to both versions it may fail at runtime, if possible nearly everything should be 1.1. That is one reason why I'm so reluctant in rushing this. This should probably not go into 18.03

vcunat commented:
For c-like stuff, if you have both in link-closure, the build will fail during linking, as there are symbols resolvable to both versions.

peti commented:
> This should probably not go into 18.03.

Actually, this PR should be merged as quickly as possible so that any remaining issues can be discovered and fixed utilizing more eyes and hands that just the 2-3 people who are working on this branch.

vcunat commented:
I'm certain that **if** it should get to 18.03, it needs to get to master soon (several days at most).

A similar thing holds for gcc-7, as we branch 18.03 in about two weeks, though there I'm confident much more is catchable during build-time.  With openssl the "dynamic" languages might give more trouble.

gleber commented:
Erlang <R20 is failing due to a configure error:
```
checking for OpenSSL kerberos 5 support... yes
checking for krb5.h in standard locations... not found
configure: WARNING: OpenSSL is configured for kerberos but no krb5.h found
```
Which disables the Erlang stdlib `crypto` library, which almost everyone uses:
```
crypto         : OpenSSL is configured for kerberos but no krb5.h found
```

gleber commented:
The best summary I've found about Erlang and OpenSSL 1.1 is here:
https://www.bountysource.com/issues/48736281-erlang-doesn-t-build-correctly-for-certain-versions-of-openssl-1-1-x

TL;DR: Erlang <R20 does not and probably will not support OpenSSL 1.1. So either we drop older Erlangs or we pin them to previous OpenSSL version.

peti commented:
So ... what do we do about this PR now? Plenty of merge conflicts have popped up over the last 2 months of inactivity and I wonder how to move this effort forward.

copumpkin commented:
It seems like a lot of the changes could be comfortably put into master without bumping the default OpenSSL version. Like adding `pkgconfig` to svn, making packages that rely strongly on OpenSSL 1.0.2 specify explicitly that in their dependencies, etc.

So perhaps easiest to split it up and put the "no regrets" changes directly into master, and then what's left is the core of changes that rely on OpenSSL 1.1 being the default?

peti commented:
@matthewbauer, huh? Why did you do that?

matthewbauer commented:
Sorry unintentional git issue.

Mic92 commented:
Archlinux did this ages ago. I am confident we can do it for 19.03!

andir commented:
I went through all the commits and rebased (or removed) them for the current `staging` state.

My current version can be found at https://github.com/andir/nixpkgs/tree/openssl-1.1

IIRC there used to be a jobset on hydra. Did I miss it?

peti commented:
> IIRC there used to be a jobset on hydra. Did I miss it?

See https://github.com/NixOS/nixpkgs/pull/22357#issue-104256005. :smile: It's at https://hydra.nixos.org/jobset/nixos/openssl-1.1.

peti commented:
@andir, I re-set this PR based on your new version. This means https://hydra.nixos.org/jobset/nixos/openssl-1.1 will be building it now.

andir commented:
@peti thanks ;)

andir commented:
The eval has been pending for about two hours. That is probably the case because that job is disabled? (And probably why I didn't find it in the first place, I guess).

![2019-03-04-135137_913x318_scrot](https://user-images.githubusercontent.com/638836/53734620-6f8eb100-3e7c-11e9-808a-8608d4f5d6b4.png)

@peti are you able to (re-)enable jobs?

andir committed:
Revert "serf: build with openssl-1.0.2 again"

This reverts commit 2c0b9ffeadecf924a4335a47f4608434f06d80b3.

vcunat commented:
"pending" means that it's queued for evaluation.  Some of our jobsets are rather expensive to evaluate (dozens of minutes).  Now it's building already.

peti commented:
I re-enabled the Hydra job (and it's completed an evaluation already, too).

andir committed:
openvswitch: use openssl 1.0.2 for the time being

andir committed:
qt48: apply openssl 1.1 patch

andir committed:
pypy: use openssl 1.0.2

While upstream claims support of OpenSSL 1.1 the test suite fails.

andir committed:
gitAndTools.git-dit: use openssl 1.0.2

There hasn't been a release in almost two years. Some (pinned)
dependency doesn't work with a newer openssl yet :/

andir committed:
libtorrent: apply openssl 1.1 patch

andir committed:
monero: lift boost override

The newer boost libraries support openssl 1.1 out of the box and
removing the override doesn't seem to break the application either.

andir commented:
Rebased on current staging & add an updated version of the qt59 patch

andir committed:
mailsend: apply openssl1.1 patch

andir committed:
openssh_hpn: pin to openssl 1.0.2

The HPN patchset doesn't yet support OpenSSL1.1. There is a potential
patch available but that seems to break during runtime [1].

[1] https://github.com/rapier1/openssh-portable/issues/14

andir committed:
partimage: apply openssl1.1 patch

The (new) patch also incorporates the now removed no-SSLv2 patch so it
has been dropped.

andir committed:
ptlib: apply opensl1.1 patch

andir committed:
golden-cheetah: 3.4 -> 3.5-DEV1903

This version bumps allows us to use a current QT5 version.

andir commented:
Rebased on current staging, added a few more patches and fixed the qt48 commit that I broke before.

andir committed:
tpm-tools: apply openssl1.1 patch

andir committed:
tpm-luks: add openssl1.1 patch

andir committed:
git-series: use openssl 1.0.2

andir committed:
globalplatform: stick with openssl 1.0.2

While upstream has patches in tree that support newer versions of
OpenSSL those haven't been released and picking them doesn't seem to be
trivial without risking breakage that I can not test. After talking to
the previous maintainer of the package this also seems like the sanest
approach for the time being.

andir commented:
Rebased one more time. Was able to remove a few of the version pinnings. 

I started another hydra eval to see if things have improved. Last time I worked on this it was pretty close to be mergeable IMHO.

globin commented:
Moved the jobset to the nixpkgs project in order to be able to compare it to staging:

https://hydra.nixos.org/jobset/nixpkgs/openssl-1.1

globin committed:
libtorrent: remove released patch

globin committed:
radsecproxy: 1.6.9 -> 1.8.0

globin committed:
realpine: remove

upstream unmaintained, doesn't build with openssl 1.1 and no recent
changes in nixpkgs, alpine from which this was forked seems to be
maintained and a superior choice.

globin committed:
vfdecrypt: remove

upstream unmaintained, doesn't build with openssl 1.1 and no recent
changes in nixpkgs

globin committed:
xml-security-c: 1.7.3 -> 2.0.2

globin committed:
xml-security-c: 1.6.3 -> 3.0.4

globin committed:
opensaml-cpp: 2.6.1 -> 3.0.1

globin committed:
shibboleth-sp: 2.6.1 -> 3.0.4.1

globin committed:
perlPackages.AuthenModAuthPubTkt: fix build

-dss1 doesn't exist for dgst in openssl 1.1, -sha1 can also handle DSA keys now

JohnAZoidberg commented on pkgs/top-level/perl-packages.nix:
```diff
@@ -687,6 +687,8 @@ let
     patchPhase = ''
       sed -i 's|my $openssl_bin = "openssl";|my $openssl_bin = "${pkgs.openssl}/bin/openssl";|' lib/Authen/ModAuthPubTkt.pm
+      # -dss1 doesn't exist for dgst in openssl 1.1, -sha1 can also handle DSA keys now
+      sed -i 's|-dss1|-sha1|' lib/Authen/ModAuthPubTkt.pm
```
I'd recommend `substituteInPlace`. That warns when upstream changes the code and the pattern no longer matches.

globin replied:
I reckon this package will never be updated again, I don't think it matters.

globin committed:
facter: remove openssl 1.0.2 pinning

globin committed:
getxbook: fix build with openssl 1.1

globin committed:
alpine: remove openssl 1.0.2 pinning

globin committed:
freeradius: remove openssl 1.0.2 pinning

globin committed:
bind: remove openssl 1.0.2 pinning

globin committed:
xml-security-c: remove openssl 1.0.2 pinning

globin committed:
sqlcipher: remove openssl 1.0.2 pinning

globin committed:
treewide: remove openssl 1.1 pinning

globin committed:
opendkim: remove openssl 1.0.2 pinning

globin committed:
qca-qt5: remove openssl 1.0.2 pinning

globin committed:
git-crypt: remove openssl 1.0.2 pinning

globin committed:
arangodb: remove openssl 1.0.2 pinning

globin committed:
all-packages: fix earlier merge

globin committed:
john: remove openssl 1.0.2 pinning

globin committed:
encfs: remove openssl 1.0.2 pinning

globin committed:
clamav: remove openssl 1.0.2 pinning

globin committed:
chunksync: remove openssl 1.0.2 pinning

globin committed:
enlightenment.efl: remove openssl 1.0.2 pinning

globin committed:
Merge remote-tracking branch 'upstream/master' into openssl-1.1

vcunat committed:
Merge branch 'master' into openssl-1.1

... to avoid the accidental gnupatch rebuild

JohnAZoidberg commented on nixos/modules/services/network-filesystems/kbfs.nix:
```diff
@@ -48,6 +48,7 @@ in {
       requires = [ "keybase.service" ];
       after = [ "keybase.service" ];
       path = [ "/run/wrappers" ];
+      unitConfig.ConditionUser = "!@system";
```
Can you please explain why?

lheckemann replied:
I think this only in here because of a funky rebase that brings in lot of unrelated commits.

lheckemann replied:
That is, this was rebased on master, which has a bunch of commits including feed9154c0ccd8b42b46cc7ddcc65e46fb08eed4 that are not on staging yet.

domenkozar commented:
Closing this one as it's completely out of date.

lheckemann commented:
@domenkozar, @globin seems to have rebased and started working on this again 10 days ago, hence the commit spam, so I don't think it should be closed. But @globin is probably the best person to know for sure :)

primeos commented:
I also don't think this should be closed.

AFAIK we don't have another PR of this kind and this change **is very important** if we want to avoid having an end-of-life version in 19.09:

https://www.openssl.org/source/:
> Note: The latest stable version is the 1.1.1 series. This is also our Long Term Support (LTS) version, supported until 11th September 2023. Our previous LTS version (1.0.2 series) will continue to be supported until 31st December 2019 (security fixes only during the last year of support). The 1.1.0 series is currently only receiving security fixes and will go out of support on 11th September 2019. All users of 1.0.2 and 1.1.0 are encouraged to upgrade to 1.1.1 as soon as possible. The 0.9.8, 1.0.0 and 1.0.1 versions are now out of support and should not be used.

Also: I though the target version of this PR was changed from 1.1.0 to 1.1.1 (or at least that there was a statement in some comment, but I didn't find it anymore :o), but looking at the current version it still seems to be based on 1.1.0 (but I might be wrong here, I only had a quick look with the web interface and this PR is too big for that).

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-19-09-feature-freeze/3707/1

saschagrunert committed:
cri-tools: 1.14.0 -> 1.15.0

- Fix go build process to include critest as well.
- Fix GitHub repo location to be `kubernetes-sigs`
- Add me as maintainer

Signed-off-by: Sascha Grunert <sgrunert@suse.com>

globin committed:
aacskeys: remove

Unmaintained, ancient and doesn't build with openssl 1.1

globin committed:
bitcoin-xt: remove

Dead project and doesn't build with openssl 1.1

globin committed:
hadoop_2_*: build with openssl 1.0.2

globin committed:
altcoins.memorycoin: remove

dead project, doesn't build with openssl 1.1

globin committed:
altcoins.pivx: mark as broken

Upstream doesn't support newer openssl versions
https://github.com/PIVX-Project/PIVX/issues/748
"Your system is most probably using openssl 1.1 which is not the
officialy supported version. Either use 1.0.1 or run again configure
with the given option."

globin committed:
python2.pkgs.sipsimple: 3.1.1 -> 3.4.2

globin committed:
altcoins.freicoin: mark as broken

doesn't build with openssl 1.1 and they use openssl 1.0.1 for testing

globin committed:
cargo-download: 0.1.1 -> 0.1.2

globin committed:
dogecoin: 1.10.0 -> 1.14.1

fixes openssl 1.1 compat

globin committed:
cutegram: remove

dead project, no openssl 1.1 and qt >= 5.9 compat

globin committed:
bitcoin-unlimited: 1.0.3.0 -> 1.6.0.1

fixes openssl 1.1 compat

globin committed:
esteidfirefoxplugin: remove

id.ee has removed it and doesn't build with openssl 1.1, there seems to
be a new plugin available

globin committed:
ettercap: 0.8.2 -> 0.8.3

fixes openssl 1.1 compat

globin committed:
golden-cheetah: build with newer qt5

primeos commented:
@globin would it be possible to give us a quick status update regarding this PR (also regarding my previous comment: https://github.com/NixOS/nixpkgs/pull/22357#issuecomment-519683341)?

My current interpretation:
- If the milestone change wasn't done by accident I'm assuming that this PR won't land in 19.09 anymore -> we'll have to maintain an end-of-life version without official security updates.
- The [Discourse post](https://discourse.nixos.org/t/nixos-19-09-feature-freeze/3707) and this PR still refer to OpenSSL 1.1 but this isn't an LTS version -> would it be possible to switch to 1.1.1 instead or would that require significantly more effort?

Is this PR still realistic (it is open for ~2.5 years now)? Could we possibly divide it into smaller PRs (e.g. updating the most essential/important packages first)? Would it help if this PR has more visibility (e.g. if we pin it on GitHub, mention it on Discourse, etc. -> maybe the community could help with testing and updating packages)? Or could some other approach work better (e.g. updating the OpenSSL version and pinging all maintainers of affected packages and packages that depend on an older OpenSSL version -> distributing this PR, e.g. like it was done here: https://github.com/NixOS/nixpkgs/pull/56265)?

In any case: Thanks for all of your effort, I really appreciate it! :) Of course this applies to anyone who worked on this PR!

(I hope this comment doesn't give the wrong impression. I'm glad that we have this PR, but I'm also a bit concerned as we're behind most other Linux distributions and this PR was now delayed multiple times while we desperately need it.)

globin committed:
Merge remote-tracking branch 'upstream/master' into openssl-1.1

globin commented:
I wasn't aware they were EOL'ing at the end of this year, I'll fix up the remaining larger issues (IIRC only Erlang) and then merge, then the rest will be fixed with ZHF. I'll also try and cherry-pick the smaller parts to other PRs to get parts in faster.

This is updating to openssl 1.1.1 anyway as openssl 1.1 changed the versioning and 1.1.1 is at least API-compatible to 1.1.0.

Also I'll base this on master as it has its own hydra jobset. (I'm also continuously coordinating with @lheckemann for 19.09)

globin committed:
Merge remote-tracking branch 'upstream/master' into openssl-1.1

globin committed:
shallot: add patches to fix with openssl 1.1

globin committed:
torque: 4.2.10 -> 6.1.3h2

globin committed:
xpwn: bump to commit including openssl 1.1 support

globin committed:
socat2pre: mark as broken

globin committed:
tw-rs: remove

Dead project, broken with openssl 1.1

globin committed:
telegram-cli: remove

Dead project, broken with openssl 1.1

globin committed:
ssvnc: remove

Dead project, broken with openssl 1.1

globin committed:
rethinkdb: mark as broken

broken with openssl 1.1

globin committed:
haskellPackages.openid: mark as broken

broken with openssl 1.1
https://github.com/elliottt/hsopenid/issues/15

globin committed:
openxpki: mark as broken

globin committed:
erlang: pin to openssl 1.0.2 for < R20

globin committed:
hadoop: fix infinite recursion

globin committed:
squid: remove v3, default to v4

Ma27 committed:
teamviewer: 14.4.2269 -> 14.5.1691

https://community.teamviewer.com/t5/Change-Logs/Linux-v14-5-1691-Full-Change-Log/m-p/69997

Also ensured that QT is wrapped properly to avoid startup issues with
different system-wide QT versions.

(cherry picked from commit 1fc1fe7e6ad86b616c16ce09b4ba7c0c6b34905a)

globin committed:
qt56: remove; unused, broken with openssl 1.1

lheckemann committed:
cheroot: 6.5.5 -> 6.5.6, fix tests with openssl 1.1

lheckemann committed:
pysocks: 1.6.8 -> 1.7.0

globin committed:
squid: fix build with openssl 1.1

globin committed:
powershell: pint to openssl 1.0.2

globin committed:
Merge remote-tracking branch 'upstream/master' into openssl-1.1

globin committed:
treewide: openssl_1_1 pinning is unnecessary

globin commented:
This is good enough to be merged to master, some minor failures in leaf packages still exist, but they can either be fixed in ZHF or marked as broken if nobody cares

globin commented:
I'll merge this to master and that back to staging/staging-next at around midday GMT if no one beats me to it.

lheckemann committed:
torque: remove commented-out sbin handling

We have a setup hook for this!

lheckemann committed:
Merge remote-tracking branch 'origin/master' into openssl-1.1

vcunat commented:
It's in staging-next as well, since 2e6bf42a2.

lheckemann commented:
#67375 for staging (WIP)

lopsided98 commented:
This causes a segfault in `rtmpsink` in `gst-plugins-bad`, because `rtmpdump` depends on 1.0 while `gst-plugins-bad` depends on 1.1. See #67842 for a fix.
